### DISCUSSION TITLE
Why is ModbusTransactionManager not in transaction.__all__?

### DISCUSSION THREAD

**Discussion by jamesbraza** [pymodbus-dev/pymodbus#1437] [Q&A] [answered]
I am writing some tests of a device I communicate with via `pymodbus`, and I want my assertions to traverse nearly the full stack.  I decided to `patch` comms at the `ModbusTransactionManager` level.

I have one question and one request:
- Why is `ModbusTransactionManager` not in `transaction.__all__`?
- Why is `__all__` at the bottom of the module?

From PEP8, `__all__` should be at the top of a module, right below the docstring: https://peps.python.org/pep-0008/#module-level-dunder-names

**Comment by janiversen** [comment, marked as answer]
If you are testing your device, you should use e.g. client_calls and not patch anything.

If you are simulating your device, I recommend using server/simulator.

But anyhow if you patch comms at ModbusTransactionManager level, you patch internally in pymodbus and at that level we do not provide __all__ (in some places you can still find it, because no code is perfect).

We consider pymodbus one module, and thus all interfaces must have __all__ to be pep8 compliant. Our internal structure is of course available in the source but it changes between version without notification.

**Reply by jamesbraza** [reply]
> you should use e.g. client_calls and not patch anything.

https://pymodbus.readthedocs.io/en/latest/search.html?q=client_calls&check_keywords=yes&area=default

Searching for `client_calls` doesn't yield anything, can you add detail there?

---

Sounds good on `pymodbus` being one module, I understand what you mean.

If I opened a PR to move the many `__all__` to the top of the module (instead of bottom), would you be open to that?

**Reply by janiversen** [reply]
If the content of __all__ is part of our API, then its welcome.

**Reply by janiversen** [reply]
examples/client_calls.py

**Comment by janiversen** [comment]
I think I misread your suggestion. The __all__ should be in __init__.py and nowhere else, and it is not logically to put it in the top, that is where we have imports etc.

What is the reason you want to move it to the top ? I might have overlooked some good arguments.

**Reply by jamesbraza** [reply]
Oh `__all__` (per PEP8) belongs at the top of a module, right below the docstring: https://peps.python.org/pep-0008/#module-level-dunder-names

It makes sense, because it's like declaring the "public API" of the module, at the top.  Currently in `pymodbus`, `__all__` is placed at the bottom of many modules.

**Reply by janiversen** [reply]
good argument, look forward to see a pull request.

**Reply by jamesbraza** [reply]
Opened https://github.com/pymodbus-dev/pymodbus/pull/1441 ✅
